### DISCUSSION TITLE
LTSP image to disk

### DISCUSSION THREAD

**Discussion by maccadia** [ltsp/ltsp#270] [Q&A]
I have been playing today with the idea of adding an IMAGE_TO_DISK option in ltsp.conf similar to IMAGE_TO_RAM but with persistence. I ended up with this "proof of concept" code (mostly copying IMAGE_TO_RAM) :

In `55-initrd-bottom.sh`
```bash
        if [ "$IMAGE_TO_RAM" = "1" ]; then
          # IMAGE_TO_RAM code here
        fi
        if [ "$IMAGE_TO_DISK" = "1" ]; then
            img=${img_src%%,*}
            rest=${img_src#$img}
            re mkdir -p /host
	    re mount -t ntfs $(readlink -f /dev/disk/by-label/Windows*) /host # assuming only one Windows labeled disk
            if [ ! $(stat -c %Y "$img") = $(stat -c %Y "/host/${img##*/}") ]; then
                clear
                echo "New image available: updating local image..."
                echo "Running: cp -a  $img /host/${img##*/}"
                re cp -a "$img" "/host/${img##*/}"
            fi
	    re umount "$rootmnt"
            img_src="/host/${img##*/}$rest"
        fi
        re mount_img_src "$img_src" "$rootmnt" "$tmpfs"
        re set_readahead "$rootmnt"
```

The LTSP image is only updated when the local image is different from the server image (different modify date timestamp).

We use LTSP to provide dual boot in our IT labs, so we have disks on our LTSP clients. The image to RAM is not suitable for us because we might end up with 5GB+ LTSP images.

I hope image to disk can have a performance benefit at the cost of loading the network sometimes. The network will never be as fast as a cheap SSD but there is probably some overhead to read from the image on ntfs.

**Comment by alkisg** [comment]
Yeah this is surely a feature that will be helpful now that SSDs are common and faster than LANs.
I'd go with MIRROR_IMAGE="ltsp-image-source".
Check https://ltsp.org/man/ltsp-ipxe/#advanced-image-sources to see an example for how an LTSP image source can mount partitions, while specifying mount options etc.

An alternative approach would be to add the LTSP image in a local GRUB entry, and boot with root=/dev/disk/by-label/Windows ltsp.image=/path/to/x86_64.img. This would allow booting without an LTSP server at all (roaming laptops etc), but it would require an additional step of installing and configuring GRUB.

**Comment by maccadia** [comment]
@alkisg thanks for the tips.

I tried to find a way to optimize the image distribution (copy) but I haven't found any. It should to be possible to replace `cp` with `rsync` using a statically linked rsync binary in intiramfs. However, I'm not sure we can expect better performances as SquashFS does not seems to be rsync friendly.

P2P might also be an option. I looked into [dat](https://docs.dat.foundation/) but it's currently not working in my network.

**Comment by alkisg** [comment]
Squashfs has options to be zsync/rsync friendly. See the rsync code in `ltsp initrd-bottom` for how to chroot to the rootfs and run rsync from there, where all the binaries etc are available.

In certain cases, it might be a better approach to boot with the old image and run rsync etc while the client is already booted and in use.

**Comment by maccadia** [comment]
> Squashfs has options to be zsync/rsync friendly

Do you know the details ? I can't find anything on google. 

I tried to rsync an LTSP SquashFS image (with default settings) and the whole file is copied : 
```bash
$ cd /srv/ltsp/images/
$ mkdir test
$ rsync -av x86_64.img test/
sending incremental file list
x86_64.img

sent 2,038,187,942 bytes  received 35 bytes  271,758,396.93 bytes/sec
total size is 2,037,690,368  speedup is 1.00
$ mv x86_64.img{.old,}
$ rsync -av x86_64.img test/
sending incremental file list
x86_64.img

sent 2,038,151,070 bytes  received 35 bytes  99,422,005.12 bytes/sec
total size is 2,037,653,504  speedup is 1.00
```

> See the rsync code in ltsp initrd-bottom for how to chroot to the rootfs and run rsync from there, where all the binaries etc are available.

I saw that, but I I don't understand it very well. The inner working of LTSP is still murky for me.

> In certain cases, it might be a better approach to boot with the old image and run rsync etc while the client is already booted and in use.

Yes, I'm thinking about it too, but it might gives errors for kernel updates.

**Comment by alkisg** [comment]
> Do you know the details ? I can't find anything on google.

I've read about it but I haven't tested it. Try reading the following URLs about zsync:
https://kparal.wordpress.com/2009/09/01/zsync-transfer-large-files-efficiently/
http://0pointer.net/blog/casync-a-tool-for-distributing-file-system-images.html
(paragraph 9. What about delivering squashfs images? How well does chunking work on compressed serializations?)

> Yes, I'm thinking about it too, but it might gives errors for kernel updates.

Ubuntu keeps the last 2 kernels installed. In a similar fashion, `ltsp kernel` could be tweaked to publish the older kernel while the clients are still updating, or the newer kernel after all updates has finished.

**Comment by maccadia** [comment]
`casync` looks promising. I did some basic tests and casync only needed to get 500 MB from the network to update a 2 GB LTSP image.

Steps : 
1. I installed apache2 and put x86_64.img and x86_64.img.old in `/var/www/html`
```bash
$ ls -l x86_64.img*
-rw-r--r-- 1 root root 2037690368 mars  12 15:39 x86_64.img
-rw-r--r-- 1 root root 2037653504 mars  12 15:39 x86_64.img.old
```
2. I created the needed files for casync to work : 
```bash
casync -v make x86_64.caibx x86_64.img
casync -v make x86_64.old.caibx x86_64.img.old
```
3. on a remote host, I downloaded x86_64.img
```bash
$ time casync -v extract  http://www.example.com/x86_64.caibx x86_64.img
real	0m44.017s
user 0m10.822s
sys	0m15.409s
```
4. again
```bash
$ time casync -v extract  http://www.example.com/x86_64.caibx x86_64.img
real	0m25.917s
user 0m22.895s
sys	0m2.351s
```
Quite some time but network is barely used.

5.  I downloaded x86_64.img.old using x86_64.img as seed
```bash
$ time casync -v extract  http://www.example.com/x86_64.old.caibx --seed=x86_64.img x86_64.img.old
Zero bytes written as sparse files: 12.4K
Bytes cloned through reflinks: 0B
Chunk requests fulfilled from local store: 0
Bytes used from local store: 0B
Chunk requests fulfilled from local seed: 25553
Bytes used from local seed: 1.4G
Chunk requests fulfilled from remote store: 5534
Bytes used from remote store: 422.8M

real	0m34.889s
user 0m24.694s
sys	0m5.244s
```

Updating took only 10s less than downloading the whole file, but most of the time is due to CPU usage and not network. So this might be useful to reduce the network usage when several LTSP clients download the image at the same time.

**Comment by maccadia** [comment]
`zsync` performs better than `casync` in my tests. With the same tests with the same files, zsync can reuse 97% (!) of the local file.

Steps :

1. (same files and web server as previous comment)

2.  I created the needed files for zsync to work :

```shell
zsyncmake -C -u http://www.example.com/x86_64.img x86_64.img
zsyncmake -C -u http://www.example.com/x86_64.img.old x86_64.img.old
```

3.  on a remote host, I downloaded x86_64.img

```shell
$ time zsync http://www.example.com/x86_64.img.zsync
#################### 100.0% 0.0 kBps DONE
 No relevent local data found
 downloading from http://www.example.com/x86_64.img:
#################### 100.0% 108338.2 kBps DONE
 verifying download...checksum matches OK
used 0 local, fetched 2037690662
real 0m21.438s
user 0m7.025s
sys 0m2.972s
```

4.  again (zsync should use 100% from local file ; indeed)

```shell
$ time zsync http://www.example.com/x86_64.img.zsync
#################### 100.0% 0.0 kBps DONE
 reading seed file x86_64.img: **********Read x86_64.img. Target 100.0% complete.
 verifying download...checksum matches OK
used 2037690368 local, fetched 0
real 0m9.977s
user 0m6.464s
sys 0m1.668s
```

5.  I downloaded x86_64.img.old using x86_64.img as seed

```shell
$ time zsync http://www.example.com/x86_64.img.old.zsync -i x86_64.img
#################### 100.0% 18.4 kBps DONE
 reading seed file x86_64.img: ***********Read x86_64.img. Target 97.7% complete.
 downloading from http://www.example.com/x86_64.img.old:
#################### 100.0% 0.0 kBps DONE
 verifying download...checksum matches OK
used 1990078464 local, fetched 48678735
real 0m12.497s
user 0m10.290s
sys 0m1.897s
```

zsync seems more efficient than casync in this context and faster overall.

**Comment by maccadia** [comment]
I managed somehow to use zsync to update the image on disk (see [55-initrd-bottom.sh](https://gist.github.com/MAccadia/1d0cbfb991bd20d2a1a31455019726c2)). But it turns out that zsync is not always efficient. I did an update on my LTSP server and rebuilt the image ; zsync could only reuse 22% of the local file and is quite slow.

**Comment by stevehol** [comment]
@MAccadia - did you arrive at a practical solution?
I'm interested mainly to reduce startup times of big apps like LibreOffice and Firefox on low power clients like HP t620 and Dell Wyse 5060 (PassMark around 2300-2500). I don't mind a slower boot.

**Comment by maccadia** [comment]
> @MAccadia - did you arrive at a practical solution?

Unfortunately no, but if you don't mind a slow boot, I would suggest to try the code I gave in the original post. I don't know however what is the "correct" way to modify 55-initrd-bottom.sh as any modification would probably be overridden by updates.

**Comment by Vinpasso** [comment]
Hi,

sorry for reviving this thread, but might it be possible to get support for something like this integrated into LTSP? It would be nice to have a central image that is stored locally for performance and memory utilization benefits.

**Reply by aleaz** [reply]
Hello everyone!

I want to share our experience with the implementation and use of LTSP over several years in a government organization. Over time, we encountered performance and scalability issues in several geographically separated units of the organization, each with a large number of users, over 200. The solution that we found most suitable was to equip the terminals (with different hardware specifications) with SSD drives to install and run the images from there, as well as to use them as local SWAP. These modifications, among others, significantly improved overall performance and user experience.

In our current architecture, terminals with and without drives coexist. Those without drives operate as usual, while those with drives, if configured correctly (with the correct LABEL), attempt to download and install or update the image if necessary in this location, and then boot from there.

Regarding the distribution and updating of the image, we evaluated various options such as rsync, scp, wget, bittorrent, etc. The most efficient option for us turned out to be online decompression of the squashfs file directly into a RAW image (uncompressed) on a disk partition.

Here are the modifications we made to the 55-initrd-bottom.sh file in case they can be of help:
```sh
if [ "$IMAGE_TO_DISK" = "1" ] && [ "$IMAGE_TO_RAM" != "1" ]; then
    # Assuming only one LTSP_ROOT labeled disk
    locdisk=$(readlink -f /dev/disk/by-label/LTSP_ROOT)
    if [ ! -z "$locdisk" ] && [ -e "/dev/disk/by-label/LTSP_ROOT" ]; then
        img=${img_src%%,*}
        re mkdir -p /host
        re mount -t ext4 "$locdisk" /host
        if [ "$IMAGE_TO_DISK_REMOVE_ALL" = "1" ]; then
            re rm -f /host/*.img /host/*.img.stat
        elif [ "$IMAGE_TO_DISK_REMOVE " = "1" ]; then
            re rm -f /host/${img##*/}*
        fi
        # Clear the screen for a better user experience :)
        re clear
        # The image is created if the file does not exist locally or if the size changed
        block_count=${IMAGE_TO_DISK_BLOCK_COUNT-15000} 
        if [ ! -f "/host/${img##*/}" ] || [ ! -f "/host/${img##*/}.stat" ] || [ $(stat -c %s "/host/${img##*/}") -ne $((1048576 * block_count)) ] ; then
            echo "Reserving space for unpacking the disk image."
            echo "Please, wait a few minutes do not turn off or unplug the computer!"
            re dd if=/dev/zero of=/host/${img##*/} bs=1048576 count=$block_count
            # To force the update
            re touch -m -t 197001010000 "/host/${img##*/}.stat"
        fi
        # The image is downloaded and unsquashed if modification date has changed;
        # otherwise, the existing image is used
        if [ ! $(stat -c %Y "$img") = $(stat -c %Y "/host/${img##*/}.stat") ]; then
            echo
            echo "New update available!."
            echo "Please, do not turn off or disconnect the device while the update is in progress."
            echo
            loopdev="$(losetup -f)"
            re losetup $loopdev /host/${img##*/}
            re mke2fs -q -t ext4 -F $loopdev > /dev/null 2>&1
            tmpdir=$(mktemp -d)
            echo "Installing update..."
            re mkdir -p $tmpdir
            re mount -o loop $loopdev $tmpdir
            re unsquashfs -q -f -d $tmpdir/ "${img}"
            re umount $tmpdir
            re losetup -d $loopdev
            re touch -m -d @$(stat -c %Y ${img}) /host/${img##*/}.stat
        else
            echo "No updates available!"
            echo "Using local image /host/${img##*/} ($(stat -c %Y /host/${img##*/}.stat))..."
        fi
        img_src="/host/${img##*/}"
        re umount "$rootmnt"
    fi
fi
```

I hope that this variant based on the previously proposed idea can help incorporate these features in future versions.

**Reply by ma-schmitz** [reply]
I have one question.
Are the thinclients still started via pxe boot or is there a bootloader(grub) on the hard disk?

**Reply by aleaz** [reply]
Hello, the thin clients continue to operate in the usual way, booting via PXE as before, with no changes in this regard.

**Comment by alkisg** [comment]
I'm using this in production to implement work-from-home PCs, but I don't know if I'll ever have enough free time to integrate it upstream in LTSP...

**Reply by ma-schmitz** [reply]
Hello alksig,

this thread is a bit older, but I am also facing the problem that I want to connect my work-from-home PC with ltsp.
Can you possibly give me some help with this?

**Comment by claudineifusieger** [comment]
Does anyone have a tutorial or video explaining how to implement ImageToDisk? I tried copying and pasting these scripts, but the client doesn't start after that.
